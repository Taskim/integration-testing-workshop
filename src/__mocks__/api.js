import uuid from 'uuidv4'

export const fetchCat = jest.fn(() => {
    const id = uuid()
    return Promise.resolve([
        {
            id,
            url: `${id}_url`,
        },
    ])
})
