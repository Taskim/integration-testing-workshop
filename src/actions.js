export const fetchCatRequest = () => ({ type: 'FETCH_CAT_REQUEST' })
export const fetchCatSuccess = cat => ({ type: 'FETCH_CAT_SUCCESS', cat })
export const fetchCatFailure = error => ({ type: 'FETCH_CAT_FAILURE', error })

export const adoptCat = () => ({ type: 'ADOPT_CAT' })
