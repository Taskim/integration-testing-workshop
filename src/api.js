export function fetchCat() {
    return fetch(
        'https://api.thecatapi.com/v1/images/search?mime_types=gif&size=med',
        {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                'x-api-key': '2aae974e-1039-408d-aff1-d839557f017a',
            },
        }
    ).then(response => response.json())
}
