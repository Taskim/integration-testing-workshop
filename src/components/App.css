.App {
    text-align: center;
    /* background: #abc3c7; */
    background-color: #e6dcdc;
    min-height: 100vh;
}

.img-container {
    height: 40vh;
    padding: 24px 0;
    display: flex;
    justify-content: center;
    align-items: center;
}

.img-container > img {
    max-height: 100%;
    width: auto;
}

.cat-button {
    border: 4px solid #f1fbf7;
    text-decoration: none;
    color: #f1fbf7;
    font-family: 'Roboto', sans-serif;
    font-size: 24px;
    padding: 20px 25px;
    transition: all 0.3s;
    background: transparent;
    cursor: pointer;
    outline: 0;
    margin: 0 24px;
}

.cat-button:hover {
    border: 4px solid #f1fbf7;
    background: url('http://catsticker.com/images/dancing-cat.gif');
    background-size: cover;
    background-position: center -40px;
    color: #f1fbf7;
}

.nope:hover {
    background: url('https://media0.giphy.com/media/rCxogJBzaeZuU/giphy.gif?cid=790b76115d47e1034f47536a7331a908&rid=giphy.gif');
}

.adopted-cats-container {
    padding: 24px;
}

.adopted-cats {
    columns: 250px auto;
}

h1 {
    margin: 96px 0 48px 0;
    text-align: left;
}

.adopted-cat {
    width: 100%;
    position: relative;
}

.adopted-cat > img {
    width: 100%;
}
