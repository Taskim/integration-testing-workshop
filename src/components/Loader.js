import React from 'react'
import './Loader.css'

const Loader = () => (
    <div className="box" data-testid="loader">
        <div className="cat">
            <div className="cat__body" />
            <div className="cat__body" />
            <div className="cat__tail" />
            <div className="cat__head" />
        </div>
    </div>
)

export default Loader
