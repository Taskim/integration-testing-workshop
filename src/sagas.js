import { takeLatest, call, put } from 'redux-saga/effects'

import { fetchCat } from './api'
import { fetchCatSuccess, fetchCatFailure } from './actions'

export function* fetchCatFlow() {
    try {
        const cat = yield call(fetchCat)
        yield put(fetchCatSuccess(cat[0]))
    } catch (e) {
        yield put(fetchCatFailure(e))
    }
}

export function* catWatcher() {
    yield takeLatest('FETCH_CAT_REQUEST', fetchCatFlow)
}
