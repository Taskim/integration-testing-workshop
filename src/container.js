import { connect } from 'react-redux'
import App from './components/App.js'
import { fetchCatRequest, adoptCat } from './actions'

export const mapStateToProps = state => ({
    cat: state.cat,
    status: state.status,
    adoptedCats: state.adoptedCats,
})

export default connect(
    mapStateToProps,
    { fetchCat: fetchCatRequest, adoptCat }
)(App)
