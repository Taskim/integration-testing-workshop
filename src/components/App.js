import React, { Component } from 'react'
import './App.css'
import Loader from './Loader.js'
import catError from './error.png'

class App extends Component {
    componentDidMount() {
        this.props.fetchCat()
    }

    adopt = () => {
        this.props.adoptCat()
        this.props.fetchCat()
    }
    render() {
        const { status, cat, fetchCat, adoptedCats } = this.props
        return (
            <div className="App" data-testid="app">
                <div className="img-container">
                    {status === 'loaded' && cat && (
                        <img
                            alt="cat"
                            src={cat.url}
                            data-testid="current-cat"
                        />
                    )}
                    {status === 'loading' && <Loader />}
                    {status === 'error' && (
                        <React.Fragment>
                            <img
                                src={catError}
                                alt="cat-error"
                                className="cat-error-img"
                                data-testid="error-img"
                            />
                            <div
                                className="error-message"
                                data-testid="error-message"
                            >
                                Déso, j'ai eu la flemme là frère. Réessaye.
                            </div>
                        </React.Fragment>
                    )}
                </div>
                <div className="button-container">
                    {status !== 'error' && (
                        <React.Fragment>
                            <button
                                disabled={status !== 'loaded'}
                                className="cat-button"
                                onClick={this.adopt}
                                data-testid="adopt-button"
                            >
                                Adopter ce chat
                            </button>
                            <button
                                className="cat-button nope"
                                onClick={fetchCat}
                                data-testid="fetch-button"
                            >
                                Voir un autre chat
                            </button>
                        </React.Fragment>
                    )}
                    {status === 'error' && (
                        <button
                            className="cat-button retry"
                            onClick={fetchCat}
                            data-testid="retry-button"
                        >
                            Réessayer
                        </button>
                    )}
                </div>
                {adoptedCats.length > 0 && (
                    <div
                        className="adopted-cats-container"
                        data-testid="collection"
                    >
                        <h1>
                            {`Mes chats`}
                            <span data-testid="adopted-cat-count">
                                {adoptedCats.length}
                            </span>
                        </h1>
                        <div className="adopted-cats">
                            {adoptedCats.map(cat => (
                                <div
                                    key={cat.id}
                                    className="adopted-cat"
                                    data-testid="adopted-cat"
                                >
                                    <img alt="adopted-cat" src={cat.url} />
                                </div>
                            ))}
                        </div>
                    </div>
                )}
            </div>
        )
    }
}

export default App
