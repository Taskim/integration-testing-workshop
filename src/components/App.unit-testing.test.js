import '@testing-library/react/cleanup-after-each'
import '@testing-library/jest-dom/extend-expect'

import React from 'react'
import { takeLatest, put, call } from 'redux-saga/effects'
import { cloneableGenerator } from '@redux-saga/testing-utils'
import { shallow } from 'enzyme'

import * as actions from '../actions'
import reducer from '../reducers'
import { catWatcher, fetchCatFlow } from '../sagas'
import { fetchCat } from '../api'
import App from './App'
import { mapStateToProps } from '../container'
import Loader from './Loader'

describe('unit testing', () => {
    const cat = {
        id: 'cat_id',
        url: 'https://cdn2.thecatapi.com/images/32t.gif',
    }
    describe('components', () => {
        it('renders the App', () => {
            expect(
                shallow(
                    <App
                        fetchCat={jest.fn()}
                        status="loading"
                        adoptedCats={[]}
                    />
                    // hack with jest@24 :shrug:
                ).debug()
            ).toMatchSnapshot()
            expect(
                shallow(
                    <App
                        fetchCat={jest.fn()}
                        cat={cat}
                        status="loaded"
                        adoptedCats={[cat]}
                    />
                ).debug()
            ).toMatchSnapshot()
        })
        it('adopts a cat', () => {
            const adoptCat = jest.fn()
            const fetchCat = jest.fn()
            const wrapper = shallow(
                <App
                    adoptCat={adoptCat}
                    fetchCat={fetchCat}
                    cat={cat}
                    status="loaded"
                    adoptedCats={[cat]}
                />
            )
            expect(fetchCat).toHaveBeenCalledTimes(1)
            expect(adoptCat).toHaveBeenCalledTimes(0)

            wrapper.instance().adopt()
            expect(adoptCat).toHaveBeenCalledTimes(1)
        })
        it('displays a spinning cat', () => {
            expect(shallow(<Loader />).debug()).toMatchSnapshot()

            const adoptCat = jest.fn()
            const fetchCat = jest.fn()

            const wrapper = shallow(
                <App
                    adoptCat={adoptCat}
                    fetchCat={fetchCat}
                    cat={null}
                    status="loading"
                    adoptedCats={[]}
                />
            )
            expect(wrapper.find('Loader')).toHaveLength(1)
        })
        it('displays a retry button in case of error', () => {
            const adoptCat = jest.fn()
            const fetchCat = jest.fn()

            expect(fetchCat).toHaveBeenCalledTimes(0)

            const wrapper = shallow(
                <App
                    adoptCat={adoptCat}
                    fetchCat={fetchCat}
                    cat={null}
                    status="error"
                    adoptedCats={[]}
                />
            )

            expect(fetchCat).toHaveBeenCalledTimes(1)

            expect(wrapper.find('.cat-error-img')).toHaveLength(1)
            expect(wrapper.find('.error-message')).toHaveLength(1)
            expect(wrapper.find('.retry')).toHaveLength(1)

            wrapper.find('.retry').simulate('click')
            expect(fetchCat).toHaveBeenCalledTimes(2)
        })
    })
    describe('actions', () => {
        it('returns a fetch cat request action 🐱', () => {
            expect(actions.fetchCatRequest()).toEqual({
                type: 'FETCH_CAT_REQUEST',
            })
        })
        it('returns a fetch cat success action', () => {
            expect(actions.fetchCatSuccess()).toEqual({
                type: 'FETCH_CAT_SUCCESS',
            })
        })
        it('returns a fetch cat failure action', () => {
            expect(actions.fetchCatFailure()).toEqual({
                type: 'FETCH_CAT_FAILURE',
            })
        })
        it('returns an adopt cat action', () => {
            expect(actions.adoptCat()).toEqual({
                type: 'ADOPT_CAT',
            })
        })
    })
    describe('reducer', () => {
        const initialState = {
            adoptedCats: [],
            cat: undefined,
            status: undefined,
        }
        it('returns the initial state', () => {
            expect(reducer(undefined, {})).toEqual(initialState)
        })
        it('updates the status for a cat request', () => {
            expect(reducer(undefined, { type: 'FETCH_CAT_REQUEST' })).toEqual({
                ...initialState,
                status: 'loading',
            })
        })
        it('sets a cat and updates the status when a cat is successfuly fetched', () => {
            expect(
                reducer(undefined, { type: 'FETCH_CAT_SUCCESS', cat })
            ).toEqual({
                ...initialState,
                status: 'loaded',
                cat,
            })
        })
        it('updates the status in case of failure', () => {
            expect(reducer(undefined, { type: 'FETCH_CAT_FAILURE' })).toEqual({
                ...initialState,
                status: 'error',
            })
        })
        it('updates the adopted cats', () => {
            const state = { ...initialState, cat }
            expect(reducer(state, { type: 'ADOPT_CAT' })).toEqual({
                ...state,
                adoptedCats: [cat],
            })
        })
    })
    describe('sagas', () => {
        const watcher = catWatcher()
        it('executes the fetch cat flow when a request is made', () => {
            expect(watcher.next().value).toEqual(
                takeLatest('FETCH_CAT_REQUEST', fetchCatFlow)
            )
        })
        it('tries to fetch a cat', () => {
            const generator = cloneableGenerator(fetchCatFlow)(
                actions.fetchCatRequest
            )
            const cat = { id: 'cat_id' }
            expect(generator.next().value).toEqual(call(fetchCat))
            expect(generator.next([cat]).value).toEqual(
                put(actions.fetchCatSuccess(cat))
            )
        })
        it('fails to fetch data', () => {
            const generator = cloneableGenerator(fetchCatFlow)(
                actions.fetchCatRequest
            )
            const error = new Error('Server Error')
            expect(generator.next().value).toEqual(call(fetchCat))
            expect(generator.throw(error).value).toEqual(
                put(actions.fetchCatFailure(error))
            )
        })
    })
    describe('api', () => {
        fetch.mockResponseOnce(JSON.stringify({ cat }))
        fetchCat().then(response => expect(response).toEqual({ cat }))
    })
    describe('container', () => {
        it('maps state to props 🤷‍♂️ and gives more coverage', () => {
            const initialState = {
                adoptedCats: [],
                cat: undefined,
                status: undefined,
            }
            expect(mapStateToProps(initialState)).toEqual({
                adoptedCats: [],
                cat: undefined,
                status: undefined,
            })
        })
    })
})
