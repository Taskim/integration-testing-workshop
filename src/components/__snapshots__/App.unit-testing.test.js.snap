// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit testing components displays a spinning cat 1`] = `
"<div className=\\"box\\" data-testid=\\"loader\\">
  <div className=\\"cat\\">
    <div className=\\"cat__body\\" />
    <div className=\\"cat__body\\" />
    <div className=\\"cat__tail\\" />
    <div className=\\"cat__head\\" />
  </div>
</div>"
`;

exports[`unit testing components renders the App 1`] = `
"<div className=\\"App\\" data-testid=\\"app\\">
  <div className=\\"img-container\\">
    <Loader />
  </div>
  <div className=\\"button-container\\">
    <button disabled={true} className=\\"cat-button\\" onClick={[Function]} data-testid=\\"adopt-button\\">
      Adopter ce chat
    </button>
    <button className=\\"cat-button nope\\" onClick={[Function: mockConstructor]} data-testid=\\"fetch-button\\">
      Voir un autre chat
    </button>
  </div>
</div>"
`;

exports[`unit testing components renders the App 2`] = `
"<div className=\\"App\\" data-testid=\\"app\\">
  <div className=\\"img-container\\">
    <img alt=\\"cat\\" src=\\"https://cdn2.thecatapi.com/images/32t.gif\\" data-testid=\\"current-cat\\" />
  </div>
  <div className=\\"button-container\\">
    <button disabled={false} className=\\"cat-button\\" onClick={[Function]} data-testid=\\"adopt-button\\">
      Adopter ce chat
    </button>
    <button className=\\"cat-button nope\\" onClick={[Function: mockConstructor]} data-testid=\\"fetch-button\\">
      Voir un autre chat
    </button>
  </div>
  <div className=\\"adopted-cats-container\\" data-testid=\\"collection\\">
    <h1>
      Mes chats
      <span data-testid=\\"adopted-cat-count\\">
        1
      </span>
    </h1>
    <div className=\\"adopted-cats\\">
      <div className=\\"adopted-cat\\" data-testid=\\"adopted-cat\\">
        <img alt=\\"adopted-cat\\" src=\\"https://cdn2.thecatapi.com/images/32t.gif\\" />
      </div>
    </div>
  </div>
</div>"
`;
